### PR TITLE
add output option: json

### PR DESCRIPTION
**Pull request by ycyaoxdu**
Signed-off-by: ycyaoxdu <yaoyuchen0626@163.com>

### CODE REVIEW

**Comment by ycyaoxdu** [comment]
```
$ clusteradm init --json
CRD successfully registered.
{
"hub-token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IkhPbVhQR1hPT1hIZDRNYUNzVEdabWhFWS0xb1l5NmdNM0dzWVBpR1RtZFkifQ.eyJhdWQiOlsiaHR0cHM6Ly9rdWJlcm5ldGVzLmRlZmF1bHQuc3ZjLmNsdXN0ZXIubG9jYWwiXSwiZXhwIjoxNjY4NTY2OTM4LCJpYXQiOjE2Njg1NjMzMzgsImlzcyI6Imh0dHBzOi8va3ViZXJuZXRlcy5kZWZhdWx0LnN2Yy5jbHVzdGVyLmxvY2FsIiwia3ViZXJuZXRlcy5pbyI6eyJuYW1lc3BhY2UiOiJvcGVuLWNsdXN0ZXItbWFuYWdlbWVudCIsInNlcnZpY2VhY2NvdW50Ijp7Im5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcCIsInVpZCI6IjZhN2ZlZDFiLTRhYjAtNDM2ZC1hOGY1LWFjNDFlMzM0NGMxOSJ9fSwibmJmIjoxNjY4NTYzMzM4LCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6b3Blbi1jbHVzdGVyLW1hbmFnZW1lbnQ6Y2x1c3Rlci1ib290c3RyYXAifQ.HovV6TYi5FYpy3ur9AWLJYHrScgVw2nj8Vh5dKOQopVvCEV0IfnZO8ZeTHwwBpiidDjRC21b0VjG90ov-6Wul-Ux7K7PSY7rjRc5Kocc8XYceZMrKLm19HDyuqky4wRV-eh2xHMWfJRP-QP3DGzZ50OJ3WM_lbQdQtr6s1tE7lFk20LK8HZSxJNPcdOWpdVqwKyX54WEhYtxGnTjkdbailoKT-Ja-ZTQXJX5uUU1742dg6BO5c03rRlVCYJ__7eE1WrW5GrPHpFf-D8k3FqPq5WX8pyiL1hYHXjjploFSPp3sqIj_8Iugd4cuVzqRFluJw18z5virOVxtQLFLKPzEQ",
"hub-apiserver": "https://127.0.0.1:52812"
}%             
                                                                                                                                                      
$ clusteradm get token --json
{
"hub-token": "eyJhbGciOiJSUzI1NiIsImtpZCI6IkhPbVhQR1hPT1hIZDRNYUNzVEdabWhFWS0xb1l5NmdNM0dzWVBpR1RtZFkifQ.eyJhdWQiOlsiaHR0cHM6Ly9rdWJlcm5ldGVzLmRlZmF1bHQuc3ZjLmNsdXN0ZXIubG9jYWwiXSwiZXhwIjoxNjY4NTY2OTQ1LCJpYXQiOjE2Njg1NjMzNDUsImlzcyI6Imh0dHBzOi8va3ViZXJuZXRlcy5kZWZhdWx0LnN2Yy5jbHVzdGVyLmxvY2FsIiwia3ViZXJuZXRlcy5pbyI6eyJuYW1lc3BhY2UiOiJvcGVuLWNsdXN0ZXItbWFuYWdlbWVudCIsInNlcnZpY2VhY2NvdW50Ijp7Im5hbWUiOiJjbHVzdGVyLWJvb3RzdHJhcCIsInVpZCI6IjZhN2ZlZDFiLTRhYjAtNDM2ZC1hOGY1LWFjNDFlMzM0NGMxOSJ9fSwibmJmIjoxNjY4NTYzMzQ1LCJzdWIiOiJzeXN0ZW06c2VydmljZWFjY291bnQ6b3Blbi1jbHVzdGVyLW1hbmFnZW1lbnQ6Y2x1c3Rlci1ib290c3RyYXAifQ.K0Txm5UXYJunLXWDZgRsjWyCyDOSN3TRfXfEmNT4DwG4jPCQfGBC8CXmpqZWxXOeKqiYYBPHOfEZkPabE6qZ-OJXU9HhTuli8DJb8YGuCaajgo4BBHH9JLP5T4q0OeMJ79P1LbQVKpuLsqpJHVaM5OvDP-M-IAqTGN93OJYso7wRv5jkHujFLIDt-lPyVoGc8cnla3rzq5yN2E2InP7qwY-3TXr6ZSqifTX_OpCTKgj3b5viuBPOBk0LjhgxJEYOeFQwKRMXKQt4-vLe1Kqki0wGJx5tTfe7G4uUnfUunrxsQ3WsXybbNuLWCzhiF8xnj9KiHxnPP47JrTX0-9m9QA",
"hub-apiserver": "https://127.0.0.1:52812"
}% 
```

**Comment by ycyaoxdu** [comment]
#287 
/assign @qiujian16 
cc @nirs

**Comment by qiujian16** [comment]
/approve
/lgtm
/hold

looks good to me in general, but would need @nirs to confirm.

**Comment by nirs** [comment]
@ycyaoxdu Thanks for the quick response!

I have some concerns about the output:

> ```
> $ clusteradm init --json
> CRD successfully registered.

I hope this is not written to stdout. The entire output should be valid
json document unless the command terminated with non zero exit code.

> {
> "hub-token": 

It would be nice to indent the output. Not required since the output if for a machine, but other tools do indent json output (e.g. kubectl).

> "eyJhbGciOiJSUzI1NiIsImtpZCI6IkhPbVhQR1hPT1hIZDRNYUNzVEdabWhFWS0xb1l5NmdNM0dzWVB...,
> "hub-apiserver": "https://127.0.0.1:52812"
> }%

I hope the `%` is not part of the output, this is not valid json.
                                                                                                                                                                  
Same questions for get token output.

File: pkg/cmd/get/token/cmd.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func NewCmd(clusteradmFlags *genericclioptionsclusteradm.ClusteradmFlags, stream
 
 	cmd.Flags().StringVar(&o.outputFile, "output-file", "", "The generated resources will be copied in the specified file")
 	cmd.Flags().BoolVar(&o.useBootstrapToken, "use-bootstrap-token", false, "If set then the bootstrap token will used instead of a service account token")
+	cmd.Flags().BoolVar(&o.jsonOutput, "json", false, "If set output will be json format")
```
Comment (nirs) [inline pkg/cmd/get/token/cmd.go:48]: This works, but what if later we want to add yaml output? I think it will be better
to have:

    --output=text (defalt)
    --output=json

This is how kubectl works, working like other tools in this space will make this
easier to use.

File: pkg/cmd/init/exec.go [hunk 2]
```diff
@@ -192,12 +195,40 @@ func (o *Options) run() error {
...
+	val := HubValue{
+		HubToken:     token,
+		HubApiserver: host,
+	}
+	b, err := json.Marshal(val)
+	if err != nil {
+		return err
+	}
+	var out bytes.Buffer
+	err = json.Indent(&out, b, "", "")
```
Comment (nirs) [inline pkg/cmd/init/exec.go:225]: Using indent string of 2 spaces makes output little nicer:
 
    json.Indent(&out, b, "", "  ")

File: pkg/cmd/get/token/exec.go [hunk 3]
```diff
@@ -4,11 +4,13 @@ package token
 import (
 	"context"
 	"fmt"
+	"os"
 
 	"github.com/spf13/cobra"
 	"github.com/stolostron/applier/pkg/apply"
 	"github.com/stolostron/applier/pkg/asset"
 	"k8s.io/apimachinery/pkg/api/errors"
+	initoption "open-cluster-management.io/clusteradm/pkg/cmd/init"
```
Comment (nirs) [inline pkg/cmd/get/token/exec.go:13]: Importing another command to share the json output works, but I think it will be cleaner to keep the code for writing json output in another package (e.g. util?)

File: pkg/cmd/init/exec.go [hunk 4]
```diff
@@ -192,12 +195,40 @@ func (o *Options) run() error {
...
+			"    %s --cluster-name <cluster_name>\n\n"+
+			"Replace <cluster_name> with a cluster name of your choice. For example, cluster1.\n\n",
+			cmd,
+		)
+	}
 
 	return apply.WriteOutput(o.outputFile, output)
 }
+
+func WriteJsonOutput(w io.Writer, token, host string) error {
```
Comment (nirs) [inline pkg/cmd/init/exec.go:215]: This works only for writing hub info, but it will be more useful it will be:

    func WriteJsonOutput(w io.Writer, v any) error {


so when we add json output to other command it can use the same helper to write json.

**Comment by ycyaoxdu** [comment]
please review again @nirs , thanks

**Comment by ycyaoxdu** [comment]
@nirs hi, could you please review changes again?

File: pkg/cmd/init/exec.go [hunk 5]
```diff
@@ -76,6 +76,11 @@ func (o *Options) validate() error {
 		return fmt.Errorf("registry should not be empty")
 	}
 
+	// If --wait is set, some information during initialize process will print to output, the output would not keep
+	// machine readable, so this behavior should be disabled
+	if o.wait && o.output != "text" {
+		return fmt.Errorf("output should be text if --wait is set")
+	}
```
Comment (nirs) [inline pkg/cmd/init/exec.go:83]: Disabling json format when using --wait is not great, but good enough.

A program can use

    clusteradm init --wait ...

To initialize, dropping or logging the text, and then use:

    clusteradm get token --output json

To fetch the hub info.

We can improve this later in several ways:
- Don't log anything progress messages when using json
- Or use jsonlines[1] format, so all lines are valid json

Possible json output:

    {"phase": "Progressing", "description": "Wait  for ... crd to be ready"}
    {"phase": "Progressing", "description": "Wait for ... crd to be established"}
    {"phase": "Progressing", "description": "CRD successfully registered"}
    {"phase: "Completed", "description": "...", "hub-token": ..., "hub-apiserver": ...}

The program reading this will read one line at a time, parsing the json. It can 
log the progress messages if it wants. It will take the hub-token and hub-apiserver
from the last message with the "Completed" phase.

In ovirt-img we use jsonlines to provide machine readable progress[2].
    
[1] https://jsonlines.org/
[2] https://github.com/oVirt/ovirt-imageio/pull/160

File: pkg/cmd/init/exec.go [hunk 6]
```diff
@@ -129,7 +134,7 @@ func (o *Options) run() error {
 	output = append(output, out...)
 
 	if !o.ClusteradmFlags.DryRun {
-		if err := helperwait.WaitUntilCRDReady(apiExtensionsClient, "clustermanagers.operator.open-cluster-management.io"); err != nil {
+		if err := helperwait.WaitUntilCRDReady(apiExtensionsClient, "clustermanagers.operator.open-cluster-management.io", o.wait); err != nil {
```
Comment (nirs) [inline pkg/cmd/init/exec.go:137]: Is this related to this change? I don't understand what this code is doing and why a wait function needs the wait flag.

File: pkg/cmd/upgrade/clustermanager/exec.go [hunk 7]
```diff
@@ -107,8 +107,8 @@ func (o *Options) run() error {
 	}
 	output = append(output, out...)
 
-	if o.wait && !o.ClusteradmFlags.DryRun {
```
Comment (nirs) [inline pkg/cmd/upgrade/clustermanager/exec.go:110]: This makes sense to me, we wait only if the wait flag is true.

File: pkg/helpers/client.go [hunk 8]
```diff
@@ -116,10 +116,12 @@ func getClusterInfoKubeConfig(kubeClient kubernetes.Interface) (*clientcmdapiv1.
...
 
 //WaitCRDToBeReady waits if a crd is ready
-func WaitCRDToBeReady(apiExtensionsClient apiextensionsclient.Interface, name string, b wait.Backoff) error {
+func WaitCRDToBeReady(apiExtensionsClient apiextensionsclient.Interface, name string, b wait.Backoff, wait bool) error {
 	errGet := retry.OnError(b, func(err error) bool {
 		if err != nil {
-			fmt.Printf("Wait  for %s crd to be ready\n", name)
+			if wait {
+				fmt.Printf("Wait  for %s crd to be ready\n", name)
+			}
```
Comment (nirs) [inline pkg/helpers/client.go:124]: So this enable the progress message only if the wait option is true. But if the wait option is false, why we we waiting for the CRD to be ready?

File: pkg/cmd/clean/exec.go [hunk 9]
```diff
@@ -79,7 +79,8 @@ func (o *Options) Run() error {
 		return nil
 	}
 	b := retry.DefaultBackoff
-	b.Duration = 1 * time.Second
+	b.Duration = 5 * time.Second
+	b.Steps = 15
```
Comment (nirs) [inline pkg/cmd/clean/exec.go:83]: Why do we need larger retry and step after the json output change? Is this unrelated fix to avoid failures in the CI?

File: pkg/cmd/unjoin/exec.go [hunk 10]
```diff
@@ -68,14 +68,15 @@ func (o *Options) run() error {
 		}
 		err = klusterletClient.OperatorV1().Klusterlets().Delete(context.Background(), "klusterlet", metav1.DeleteOptions{})
 		if errors.IsNotFound(err) {
-			fmt.Fprintf(o.Streams.Out, "klusterlet is cleaned up already")
+			fmt.Fprintf(o.Streams.Out, "klusterlet is cleaned up already\n")
```
Comment (nirs) [inline pkg/cmd/unjoin/exec.go:71]: Unrelated fix?

File: test/e2e/e2e-test.mk [hunk 11]
```diff
@@ -24,9 +24,9 @@ start-cluster:
 .PHONY: start-cluster 
 
 test-e2e: clean-e2e ensure-kubebuilder-tools start-cluster deps install
-	go test -v ./test/e2e/clusteradm --timeout 1800s
+	go test -v ./test/e2e/clusteradm --timeout 3600s
```
Comment (nirs) [inline test/e2e/e2e-test.mk:27]: Unrelated fix?

**Comment by nirs** [comment]
@ycyaoxdu thanks for this change!

The new --output option looks good but the json module can be simplified

I don't understand the other changes (retries, timeouts), looks like unrelated fixes to me.

File: pkg/helpers/client.go (hunk 8 above)
Comment (ycyaoxdu) [reply pkg/helpers/client.go:124]: we should always wait for CRD ready to avoid [Race Condition](https://github.com/open-cluster-management-io/clusteradm/issues/47), so we kept the function name Wait in function name.
but I think the output should not be print to stdout while `--output json` is set, so I passed wait flag to it to check whether print the waiting status.

File: pkg/cmd/upgrade/clustermanager/exec.go (hunk 7 above)
Comment (ycyaoxdu) [reply pkg/cmd/upgrade/clustermanager/exec.go:110]: actually this is bug fixed before, to avoid [Race Condition](https://github.com/open-cluster-management-io/clusteradm/issues/47)

File: pkg/cmd/clean/exec.go (hunk 9 above)
Comment (ycyaoxdu) [reply pkg/cmd/clean/exec.go:83]: yes, this is related to avoid failures in the CI.

File: test/e2e/e2e-test.mk (hunk 11 above)
Comment (ycyaoxdu) [reply test/e2e/e2e-test.mk:27]: set a larger timeout for e2e test

**Comment by ycyaoxdu** [comment]
Hi @nirs could you please take a look on this pr again? I seperated the json feature and the e2e-test fix into 2 commits, maybe you can just focus on the json feature one: https://github.com/open-cluster-management-io/clusteradm/pull/288/commits/ad0d40f80450114047fe32b4394e4e5e68160764 
 and for now in init command, if --output is set to json, --wait is allowed but not print the progress information.

**Review by nirs — APPROVED** [review]
Reviewed the first commit adding --output=json, looks good now.

**Comment by nirs** [comment]
I also tested this PR with this ramen PR:
https://github.com/RamenDR/ramen/pull/638

**Comment by ycyaoxdu** [comment]
> I also tested this PR with this ramen PR: [RamenDR/ramen#638](https://github.com/RamenDR/ramen/pull/638)
> 
> @ycyaoxdu do you have an idea when this change will be released?

@nirs , hi nir, sorry for the delay, now this change is merged.

**Comment by nirs** [comment]
@ycyaoxdu Thanks!

I want to consume this change from the next release. Looking at the [releases page](https://github.com/open-cluster-management-io/clusteradm/releases), can we have
a release in December?

**Comment by ycyaoxdu** [comment]
> @ycyaoxdu Thanks!
> 
> I want to consume this change from the next release. Looking at the [releases page](https://github.com/open-cluster-management-io/clusteradm/releases), can we have a release in December?

Hi @nirs, So sorry to telling you the next release would in late January. If the release date is too late for you, you might ask for this change in our [community slack channel](https://kubernetes.slack.com/channels/open-cluster-mgmt).

**Comment by nirs** [comment]
@ycyaoxdu late January is fine, thanks!

**Comment by nirs** [comment]
@ycyaoxdu late January is now. I guess the release is expected in the next week?

**Comment by ycyaoxdu** [comment]
@nirs Hi, sorry for the delay, clusteradm V0.5.0 is available. 
Clusteradm always publish new releases after other OCM main components' new releases published. This is the reason for the delay.

**Comment by nirs** [comment]
@ycyaoxdu thanks!
